File: src/arch/x86_64/gdt.zig
// Arcturus - Hobbyist operating system written in Zig.
// Copyright (C) 2025 Theomund
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

const instruction = @import("instruction.zig");
const register = @import("register.zig");
const std = @import("std");
const tss = @import("tss.zig");

const Access = packed struct {
    accessed: u1,
    read_write: u1,
    direction_conforming: u1,
    executable: u1,
    descriptor_type: u1,
    descriptor_privilege: u2,
    present: u1,
};

pub const Flags = packed struct {
    reserved: u1,
    long_mode: u1,
    size: u1,
    granularity: u1,
};

const Descriptor = packed struct {
    limit_low: u16,
    base_low: u24,
    access: Access,
    limit_high: u4,
    flags: Flags,
    base_high: u8,
};

pub const Pointer = packed struct {
    limit: u16,
    base: u64,
};

const Selector = packed struct {
    requested_privilege_level: u2,
    table_indicator: u1,
    index: u13,
};

pub const Table = struct {
    descriptors: [7]u64,
    selectors: [6]u16,

    pub fn init(task_state_segment: *const tss.Segment) Table {
        const null_descriptor = Descriptor{
            .limit_low = 0x0,
            .base_low = 0x0,
            .access = .{
                .accessed = 0,
                .read_write = 0,
                .direction_conforming = 0,
                .executable = 0,
                .descriptor_type = 0,
                .descriptor_privilege = 0,
                .present = 0,
            },
            .limit_high = 0x0,
            .flags = .{
                .reserved = 0,
                .long_mode = 0,
                .size = 0,
                .granularity = 0,
            },
            .base_high = 0x0,
        };

        const null_selector = Selector{
            .requested_privilege_level = 0,
            .table_indicator = 0,
            .index = 0,
        };

        const kernel_code_descriptor = Descriptor{
            .limit_low = 0xFFFF,
            .base_low = 0x0,
            .access = .{
                .accessed = 0,
                .read_write = 1,
                .direction_conforming = 0,
                .executable = 1,
                .descriptor_type = 1,
                .descriptor_privilege = 0,
                .present = 1,
            },
            .limit_high = 0xF,
            .flags = .{
                .reserved = 0,
                .long_mode = 1,
                .size = 0,
                .granularity = 1,
            },
            .base_high = 0x0,
        };

        const kernel_code_selector = Selector{
            .requested_privilege_level = 0,
            .table_indicator = 0,
            .index = 1,
        };

        const kernel_data_descriptor = Descriptor{
            .limit_low = 0xFFFF,
            .base_low = 0x0,
            .access = .{
                .accessed = 0,
                .read_write = 1,
                .direction_conforming = 0,
                .executable = 0,
                .descriptor_type = 1,
                .descriptor_privilege = 0,
                .present = 1,
            },
            .limit_high = 0xF,
            .flags = .{
                .reserved = 0,
                .long_mode = 0,
                .size = 1,
                .granularity = 1,
            },
            .base_high = 0x0,
        };

        const kernel_data_selector = Selector{
            .requested_privilege_level = 0,
            .table_indicator = 0,
            .index = 2,
        };

        const user_code_descriptor = Descriptor{
            .limit_low = 0xFFFF,
            .base_low = 0x0,
            .access = .{
                .accessed = 0,
                .read_write = 1,
                .direction_conforming = 0,
                .executable = 1,
                .descriptor_type = 1,
                .descriptor_privilege = 3,
                .present = 1,
            },
            .limit_high = 0xF,
            .flags = .{
                .reserved = 0,
                .long_mode = 1,
                .size = 0,
                .granularity = 1,
            },
            .base_high = 0x0,
        };

        const user_code_selector = Selector{
            .requested_privilege_level = 3,
            .table_indicator = 0,
            .index = 3,
        };

        const user_data_descriptor = Descriptor{
            .limit_low = 0xFFFF,
            .base_low = 0x0,
            .access = .{
                .accessed = 0,
                .read_write = 1,
                .direction_conforming = 0,
                .executable = 0,
                .descriptor_type = 1,
                .descriptor_privilege = 3,
                .present = 1,
            },
            .limit_high = 0xF,
            .flags = .{
                .reserved = 0,
                .long_mode = 0,
                .size = 1,
                .granularity = 1,
            },
            .base_high = 0x0,
        };

        const user_data_selector = Selector{
            .requested_privilege_level = 3,
            .table_indicator = 0,
            .index = 4,
        };

        const task_state_base = @intFromPtr(task_state_segment);

        const task_state_limit = @sizeOf(tss.Segment) - 1;

        const task_state_descriptor = tss.Descriptor{
            .low = .{
                .limit_low = @truncate(task_state_limit),
                .base_low = @truncate(task_state_base),
                .access = .{
                    .segment_type = 0x9,
                    .descriptor_type = 0,
                    .descriptor_privilege = 0,
                    .present = 1,
                },
                .limit_high = @truncate(task_state_limit >> 16),
                .flags = .{
                    .reserved = 0,
                    .long_mode = 0,
                    .size = 0,
                    .granularity = 0,
                },
                .base_middle = @truncate(task_state_base >> 24),
            },
            .high = .{
                .base_high = @truncate(task_state_base >> 32),
                .reserved = 0,
            },
        };

        const task_state_selector = Selector{
            .requested_privilege_level = 0,
            .table_indicator = 0,
            .index = 5,
        };

        return Table{
            .descriptors = .{
                @bitCast(null_descriptor),
                @bitCast(kernel_code_descriptor),
                @bitCast(kernel_data_descriptor),
                @bitCast(user_code_descriptor),
                @bitCast(user_data_descriptor),
                @bitCast(task_state_descriptor.low),
                @bitCast(task_state_descriptor.high),
            },
            .selectors = .{
                @bitCast(null_selector),
                @bitCast(kernel_code_selector),
                @bitCast(kernel_data_selector),
                @bitCast(user_code_selector),
                @bitCast(user_data_selector),
                @bitCast(task_state_selector),
            },
        };
    }

    pub fn pointer(self: Table) Pointer {
        return Pointer{
            .limit = (self.descriptors.len * @sizeOf(u64)) - 1,
            .base = @intFromPtr(&self.descriptors[0]),
        };
    }

    pub fn load(self: Table) void {
        const ptr = self.pointer();
        instruction.cli();
        instruction.lgdt(&ptr);
        register.CS.set(self.selectors[1]);
        register.DS.set(self.selectors[2]);
        register.ES.set(self.selectors[2]);
        register.FS.set(self.selectors[2]);
        register.GS.set(self.selectors[2]);
        register.SS.set(self.selectors[2]);
    }
};

test "Pointer Limit" {
    const segment = tss.Segment.init();
    const table = Table.init(&segment);
    const pointer = table.pointer();
    try std.testing.expectEqual(0x37, pointer.limit);
}

test "Null Segment Descriptor" {
    const segment = tss.Segment.init();
    const table = Table.init(&segment);
    try std.testing.expectEqual(0x0, table.descriptors[0]);
}

test "Null Segment Selector" {
    const segment = tss.Segment.init();
    const table = Table.init(&segment);
    try std.testing.expectEqual(0x0, table.selectors[0]);
}

test "Kernel Code Segment Descriptor" {
    const segment = tss.Segment.init();
    const table = Table.init(&segment);
    try std.testing.expectEqual(0x00AF9A000000FFFF, table.descriptors[1]);
}

test "Kernel Code Segment Selector" {
    const segment = tss.Segment.init();
    const table = Table.init(&segment);
    try std.testing.expectEqual(0x8, table.selectors[1]);
}

test "Kernel Data Segment Descriptor" {
    const segment = tss.Segment.init();
    const table = Table.init(&segment);
    try std.testing.expectEqual(0x00CF92000000FFFF, table.descriptors[2]);
}

test "Kernel Data Segment Selector" {
    const segment = tss.Segment.init();
    const table = Table.init(&segment);
    try std.testing.expectEqual(0x10, table.selectors[2]);
}

test "User Code Segment Descriptor" {
    const segment = tss.Segment.init();
    const table = Table.init(&segment);
    try std.testing.expectEqual(0x00AFFA000000FFFF, table.descriptors[3]);
}

test "User Code Segment Selector" {
    const segment = tss.Segment.init();
    const table = Table.init(&segment);
    try std.testing.expectEqual(0x1B, table.selectors[3]);
}

test "User Data Segment Descriptor" {
    const segment = tss.Segment.init();
    const table = Table.init(&segment);
    try std.testing.expectEqual(0x00CFF2000000FFFF, table.descriptors[4]);
}

test "User Data Segment Selector" {
    const segment = tss.Segment.init();
    const table = Table.init(&segment);
    try std.testing.expectEqual(0x23, table.selectors[4]);
}

test "Task State Segment Selector" {
    const segment = tss.Segment.init();
    const table = Table.init(&segment);
    try std.testing.expectEqual(0x28, table.selectors[5]);
}

File: architecture/src/x86_64/gdt.rs
// Arcturus - Hobbyist operating system written in Rust.
// Copyright (C) 2025 Theomund
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

use super::instruction;
use super::register;
use super::tss::Segment;

pub struct Table {
    descriptors: [Descriptor; 7],
    selectors: [Selector; 6],
}

#[repr(C, packed(2))]
pub struct Register {
    limit: u16,
    base: u64,
}

#[derive(Copy, Clone)]
#[repr(transparent)]
pub struct Descriptor(pub u64);

impl Descriptor {
    fn new(base: u32, limit: u32, access: u8, flags: u8) -> Self {
        let limit_low = u64::from(limit & 0xFFFF);
        let base_low = u64::from(base & 0xFFFF);
        let base_middle = u64::from((base >> 16) & 0xFF);
        let access_low = u64::from(access);
        let limit_high = u64::from((limit >> 16) & 0xF);
        let flags_low = u64::from(flags);
        let base_high = u64::from((base >> 24) & 0xFF);

        let value = limit_low
            | (base_low << 16)
            | (base_middle << 32)
            | (access_low << 40)
            | (limit_high << 48)
            | (flags_low << 52)
            | (base_high << 56);

        Self(value)
    }

    fn new_system(base: u32) -> Self {
        let value = u64::from(base);

        Self(value)
    }
}

#[derive(Copy, Clone)]
#[repr(transparent)]
pub struct Selector(pub u16);

impl Selector {
    #[must_use]
    fn new(index: u16, privilege_level: u16) -> Self {
        Self((index << 3) | privilege_level)
    }
}

impl Table {
    #[must_use]
    pub fn new(segment: *const Segment) -> Self {
        let null_descriptor = Descriptor::new(0x0, 0x0, 0x0, 0x0);
        let null_selector = Selector::new(0, 0);

        let kernel_code_descriptor = Descriptor::new(0x0, 0xFFFFF, 0x9A, 0xA);
        let kernel_code_selector = Selector::new(1, 0);

        let kernel_data_descriptor = Descriptor::new(0x0, 0xFFFFF, 0x92, 0xC);
        let kernel_data_selector = Selector::new(2, 0);

        let user_code_descriptor = Descriptor::new(0x0, 0xFFFFF, 0xFA, 0xA);
        let user_code_selector = Selector::new(3, 3);

        let user_data_descriptor = Descriptor::new(0x0, 0xFFFFF, 0xF2, 0xC);
        let user_data_selector = Selector::new(4, 3);

        let task_state_base = segment as u64;
        let task_state_base_low = (task_state_base & 0xFFFF_FFFF) as u32;
        let task_state_base_high = (task_state_base >> 32) as u32;
        let task_state_limit =
            u32::try_from(size_of::<Segment>() - 1).expect("Failed to calculate limit.");

        let task_state_descriptor_low =
            Descriptor::new(task_state_base_low, task_state_limit, 0x89, 0x0);
        let task_state_descriptor_high = Descriptor::new_system(task_state_base_high);
        let task_state_selector = Selector::new(5, 0);

        Self {
            descriptors: [
                null_descriptor,
                kernel_code_descriptor,
                kernel_data_descriptor,
                user_code_descriptor,
                user_data_descriptor,
                task_state_descriptor_low,
                task_state_descriptor_high,
            ],
            selectors: [
                null_selector,
                kernel_code_selector,
                kernel_data_selector,
                user_code_selector,
                user_data_selector,
                task_state_selector,
            ],
        }
    }

    fn base(&self) -> u64 {
        self.descriptors.as_ptr() as u64
    }

    #[must_use]
    pub fn descriptor(&self, index: usize) -> Descriptor {
        self.descriptors[index]
    }

    fn limit(&self) -> u16 {
        (size_of_val(&self.descriptors) - 1) as u16
    }

    #[must_use]
    pub fn register(&self) -> Register {
        Register {
            limit: self.limit(),
            base: self.base(),
        }
    }

    #[must_use]
    pub fn selector(&self, index: usize) -> Selector {
        self.selectors[index]
    }

    pub fn load(&self) {
        instruction::cli();

        let register = self.register();
        instruction::lgdt(&register);

        register::CS::set(self.selector(1));
        register::DS::set(self.selector(2));
        register::ES::set(self.selector(2));
        register::FS::set(self.selector(2));
        register::GS::set(self.selector(2));
        register::SS::set(self.selector(2));
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn create_table() -> Table {
        let segment = Segment::new();
        Table::new(&segment)
    }

    fn get_descriptor(index: usize) -> u64 {
        let table = create_table();
        table.descriptor(index).0
    }

    fn get_selector(index: usize) -> u16 {
        let table = create_table();
        table.selector(index).0
    }

    #[test]
    fn test_null_descriptor() {
        assert_eq!(get_descriptor(0), 0x0000000000000000);
    }

    #[test]
    fn test_null_selector() {
        assert_eq!(get_selector(0), 0x0000);
    }

    #[test]
    fn test_kernel_code_descriptor() {
        assert_eq!(get_descriptor(1), 0x00AF9A000000FFFF);
    }

    #[test]
    fn test_kernel_code_selector() {
        assert_eq!(get_selector(1), 0x0008);
    }

    #[test]
    fn test_kernel_data_descriptor() {
        assert_eq!(get_descriptor(2), 0x00CF92000000FFFF);
    }

    #[test]
    fn test_kernel_data_selector() {
        assert_eq!(get_selector(2), 0x0010);
    }

    #[test]
    fn test_user_code_descriptor() {
        assert_eq!(get_descriptor(3), 0x00AFFA000000FFFF);
    }

    #[test]
    fn test_user_code_selector() {
        assert_eq!(get_selector(3), 0x001B);
    }

    #[test]
    fn test_user_data_descriptor() {
        assert_eq!(get_descriptor(4), 0x00CFF2000000FFFF);
    }

    #[test]
    fn test_user_data_selector() {
        assert_eq!(get_selector(4), 0x0023);
    }

    #[test]
    fn test_task_state_descriptor() {
        assert_ne!(get_descriptor(5), 0x0000000000000000);
        assert_ne!(get_descriptor(6), 0x0000000000000000);
    }

    #[test]
    fn test_task_state_selector() {
        assert_eq!(get_selector(5), 0x0028);
    }
}
